### ISSUE TITLE
[resize-observer] Which box information should we pass to the callback

### ISSUE THREAD

**Issue by gregwhitworth** [w3c/csswg-drafts#3329]
We now allow authors to determine which box(es) they wish to observe on layout changes. Now we need to discuss which boxes we provide the information on. We have a few options:

Take the following example: If the author sets `content-box` and `border-box` to be observed, which box dimensions would you expect to receive passed to your callback?

1. The box(es) that were requested to be observed: so in the example above, `content-box` and `border-box` information
2. They receive all available CSS boxes that are observable: They will receive `border-box`, `content-box`, `scroll-box` dimensions.

**Option 1 Pros** is good because it makes sense that I requested a specific set of boxes to be observed and those are the ones that you returned information on.

**Option 1 Cons** If the sets the observer to the `border-box` but desires to know something about an inner box, they may request it in some other manner causing us to do additional work which defeats the purpose of this API altogether.

**Option 2 Pros** You can set one observation and get information about all of the boxes, so if you have a scenario in which you want to observe the border box but check the content or scroll box - this is possible without having to observe them.

**Option 2 Cons** It clutters up the response, if I only care about the `border-box` but I get results for every other box this may be a nuisance.

My personal opinion is to go with option 2 as I think people will either over observe to get the potential rects they care about or will ask for them in some other manner. That said, I don't have a strong opinion about this as I think it will be user preference. I have a [twitter poll](https://twitter.com/gregwhitworth/status/1062954297757904898) going about this in hopes of crowd sourcing some thoughts from web devs.

**Comment by atotic** [comment]
My vote is also for #2, report all boxes. 

Observing content-box and border-box at the same time increases likelihood of "undelivered notifications error". We do not want developers observing multiple sizes just to be able to read their values out of the ResizeObserverEntry.

**Comment by gregwhitworth** [comment]
> increases likelihood of "undelivered notifications error"

Then what is the reasoning behind allowing a sequence of options to be passed in? Maybe we should limit it to only one option being passed in as to avoid this issue?

**Comment by atotic** [comment]
> Then what is the reasoning behind allowing a sequence of options to be passed in? Maybe we should limit it to only one option being passed in as to avoid this issue?

Good point. I think single option will be vast majority of usage. For the rare case of multiple sizes being observed, developers can use multiple ResizeObservers.

**Comment by gregwhitworth** [comment]
> Good point. I think single option will be vast majority of usage. For the rare case of multiple sizes being observed, developers can use multiple ResizeObservers.

To be clear here - you're in agreement that we should change this:
```
dictionary ResizeObserverOptions {
    (ResizeObserverSizeOptions or sequence<ResizeObserverSizeOptions>) size = "content-box";
};
```
to this:
```
dictionary ResizeObserverOptions {
    (ResizeObserverSizeOptions) size = "content-box";
};
```

**Comment by atotic** [comment]
Yes :-)

I started off against the change. Then I wrote down my arguments for allowing multiples, and they were not convincing. So I changed my mind, and now agree with you.

**Comment by gregwhitworth** [comment]
To summarize, the change that @atotic and I made is to be able to observe and single box at a time with all boxes being returned on callback.

**Comment by gregwhitworth** [comment]
To be very clear, here are the boxes whos dimensions would be returned:

- borderBoxSize
- contentSize
- scrollSize
- devicePixelBorderBoxSize

**Comment by TremayneChrist** [comment]
> To be very clear, here are the boxes whos dimensions would be returned:
> 
> * borderBoxSize
> * contentSize
> * scrollSize
> * devicePixelBorderBoxSize

If we can observe `border-box`, `content-box`, `scroll-box` and `device-pixel-border-box`, shouldn't all property names also include __*box*__ in the name?
 * borderBoxSize
 * contentBoxSize
 * scrollBoxSize
* devicePixelBorderBoxSize

**Comment by bfgeek** [comment]
My feedback on the call was something like:
```js
ro.observe(elem, {box: ['content-box', 'border-box']});
ro.observe(elem2, {box: 'content-box'});

function callback(entry) {
  // elem2
  entry.borderBoxSize === null; // This entry null'd out as not explicitly observed.
}
```

**Comment by gregwhitworth** [comment]
@bfgeek There would need to be two different option sets, because we're trying to keep the API shape simple - what you have defined right now would create two tasks of the `content-box` and `border-box` which, as noted by @atotic in this [comment](https://github.com/w3c/csswg-drafts/issues/3329#issuecomment-439360802) that this can lead to missing notifications. In order to achieve what you want and avoid the missing notifications is to do something like this:

    ro.observe(elem, {
                      boxToObserve: 'content-box', 
                      boxesToReturnResultsFor: ['border-box', 'content-box']
                      });

This would allow the author to trigger on a specific box's dimension changing and which dimensions the author wants returned.

**Comment by gregwhitworth** [comment]
@TremayneChrist Thanks - Opened a seperate issue to track that

**Comment by TremayneChrist** [comment]
So now we would have something like this?

```js
const ro = new ResizeObserver(entries => {
  entries.forEach(entry => {
    console.log(entry);
    // { target: Element, size: ResizeObserverSize }
  })
})
```

**Comment by atotic** [comment]
> So now we would have something like this?
>     // { target: Element, size: ResizeObserverSize }

No. Our current thinking is to keep all reported sizes in ResizeObserverEntry, but only reported one would be non-null.

This reminds me of an exception I forgot to mention in my previous comment.

"Only one size can be safely used at a time" rule has an exception. If "device-pixel-border-box" is being watched, "border-box" size is useful, and can be used safely. 

Example: in response to "device-pixel-border-box" size notification, <canvas> Element would be resized to "border-box" size, and canvas context would be resized to "device-pixel-border-box".

**Comment by gregwhitworth** [comment]
Ok, so after discussing this at length with @atotic and taking into account the valid feedback from @bfgeek regarding stale information we've come up with two potential shapes this API can take. The question comes down to:

**Do we only return the information for the box being observed, or do we return all potentially observable boxes but only fill in layout information of the box actively observed?**

**Option A:**
```
entry = {
        target: <Element>,
        contentRect: {/* v1 for back compat */},
        inline: 400,
        block: 400
 }
```
**Option B:**
```
entry = {
        target: <Element>,
        contentRect: {/* v1 for back compat */},
        contentBox: null,
        borderBox: {
            inline: 400,
            block: 400
        },
        scrollBox: null
    }
```

_Note: This does not show offset information by design as that will be discussed in issue #3550 - if we resolve to add the offsets those will go alongside `inline` & `block`, no matter which shape we decide on._

_Note 2: How fragmentation will be supported is being looked at in issue #3673 and we will adjust the shape further based on the outcome of that discussion_

**Comment by TremayneChrist** [comment]
Hey guys, just thought I'd share my initial feelings on both options.

### Option A

Pros:
1. Looks cleaner.
2. Properties can always be expected and have value.
3. Less nesting of values.
4. Reduces complexity of callback handlers.

Cons:
1. Confusion on which box is being returned.
2. Potentially less flexible.
3. Could restrict API extension in the future, or, start to clutter the global scope of the entry.

### Option B

Pros:
1. Value is tied to the observed box, reducing confusion.
2. Potentially gives more flexibility for libraries using the API.

Cons:
1. Gives the impression that you can observe multiple boxes, using the same observer. I assume not, but maybe you can?
2. Potentially adds more logic to the observer callback, for value checks.

**Comment by TremayneChrist** [comment]
Also, just to clarify...

1. So `devicePixelBorderBox` is being scrapped?

2. Sticking to the term `box` instead of `size`? I guess this is to support offsets if required (#3550)?

3. What happens if this occurs?
```js
const ro = new ResizeObserver(() => {});

ro.observe(document.body, { box: 'border-box' });

// This one will be skipped?
// or will contentBox and borderBox be watched and returned?
ro.observe(document.body, { box: 'content-box' });
```

**Comment by gregwhitworth** [comment]
@TremayneChrist We're going to be discussing all of these options today in a breakout session at the CSSWG meeting - all of them will hopefully lead to resolutions for each that are very clear. We'll have clear resolutions (if we get them) by tomorrow EOD in the minutes that get appended via the IRC bot.

> What happens if this occurs?

Regarding your current question - that's a solid one but I would expect it to behave the same as other observers which effectively the latest observer's config. Using mutation observer as an example:

https://jsbin.com/cogusojiwo/edit?html,js,console,output

Here you can see that I am both adding a new element and adding an attribute but the callback only fires on the attribute. So I would expect RO to behave the same way and you would only get an observation of the `content-box` here. Also, you're not going to want to watch numerous boxes of the same element, which is why we removed the capability to pass in an array of boxes to observer as it can lead to an increase of missed events. @atotic just in case RO behaves differently from MutationObserver for some reason.

**Comment by gregwhitworth** [comment]
To summarize the above - is to go with Option B with the removal of scrollBox - and taking into account fragments of course - so the shape will now look like this:

```
entry = {
        target: <Element>,
        contentRect: {/* v1 for back compat */},
        contentBox: null,
        borderBox: [{
            inline: 400,
            block: 400
        }]
    }
```
